### PR TITLE
[Unit Test] Make the nesting logic for vstest and NUnit(Mono) consistent

### PR DESCRIPTION
**Pull request by ketttY**
Nesting:
**Solution -> Project -> Namespace -> Test class -> Test case**
Screen Shot 2020-01-02 at 10 12 54 AM

Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1042673/

### CODE REVIEW

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs [hunk 1]
```diff
@@ -268,16 +268,28 @@ void FillTests (NunitTestInfo ti)
...
-				if (test.Tests != null)
-					newTest = new NUnitTestSuite (this, test);
-				else
-					newTest = new NUnitTestCase (this, test, test.PathName);
+				var newTest = new NUnitTestSuite (this, test);
 				newTest.FixtureTypeName = test.FixtureTypeName;
 				newTest.FixtureTypeNamespace = test.FixtureTypeNamespace;
 				Tests.Add (newTest);
 
+				Queue queue = new Queue ();
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs]:
```suggestion
				var queue = new Queue<NUnitTestSuite> ();
```

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs [hunk 2]
```diff
@@ -268,16 +268,28 @@ void FillTests (NunitTestInfo ti)
...
+				queue.Enqueue (newTest);
+
+				var forceLoad = newTest.Tests;
+				while (queue.Count != 0) {
+					NUnitTestSuite unhandledList = (NUnitTestSuite) queue.Dequeue ();
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs]:
```suggestion
					NUnitTestSuite unhandledList = queue.Dequeue ();
```

**Comment by ketttY** [comment]
@monojenkins rebase

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs [hunk 3]
```diff
@@ -268,16 +268,28 @@ void FillTests (NunitTestInfo ti)
...
+				var queue = new Queue<NUnitTestSuite> ();
+				queue.Enqueue (newTest);
+
+				var forceLoad = newTest.Tests;
+				while (queue.Count != 0) {
+					NUnitTestSuite unhandledList = queue.Dequeue ();
+					foreach (NUnitTestSuite unhandeledTest in unhandledList.UnhandledTests) {
+						Tests.Add (unhandeledTest);
+					}
+					foreach (UnitTest unhandeledTest in unhandledList.Tests) {
```
Comment (mrward) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs]: Typo here and above.

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs [hunk 4]
```diff
@@ -37,6 +37,8 @@ class NUnitTestSuite: UnitTestGroup
 		NUnitAssemblyTestSuite rootSuite;
 		string fullName;
 		
+		public UnitTestCollection UnhandledTests;
+
```
Comment (mrward) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs]: Probably should be a property.

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs [hunk 5]
```diff
@@ -80,16 +83,32 @@ protected override void OnCreateTests ()
...
+				if (test.Tests != null) {
+					foreach (NunitTestInfo info in test.Tests) {
+						if (info.Tests != null) {
+							isNameSpace = true;
+						}
+					}
+				}
+
+				if (isNameSpace) {
+					((NUnitTestSuite)newTest).testInfo.Name = ((NUnitTestSuite)newTest).testInfo.TestId;
```
Comment (mrward) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs]: Can we do this cast once? Or store the NUnitTestSuite when it is created?

**Comment by ketttY** [comment]
@monojenkins rebase

**Comment by mrward** [comment]
Seeing odd nesting.

Screenshot 2020-01-07 at 12 19 52

This is with a single test fixture in a MyProject.Tests namespace for a NUnit library project.

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs [hunk 6]
```diff
@@ -60,6 +63,12 @@ public NUnitTestSuite (NUnitAssemblyTestSuite rootSuite, NunitTestInfo tinfo): b
 				return true;
 			}
 		}
+
+		public UnitTestCollection UnhandledTests {
```
Comment (mrward) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs]: Just wondering if there is a better name for this. They seem to be a set of Namespace Test Suites and not really unhandled tests. Not sure if there is a better name here indicating they are namespaces?

**Comment by ketttY** [comment]
Single test fixture in a MyProject.Tests namespace:
test_oneNamespace

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs [hunk 7]
```diff
@@ -80,16 +81,46 @@ protected override void OnCreateTests ()
...
+					foreach (NunitTestInfo child in test.Tests) {
+						if (child.Tests != null && child.Tests [0].Tests == null) {
+							hasClassAsChild = true;
+						}
+					}
+
+					if (isNameSpace) {
+						var forceLoad = newTest.Tests;
+						foreach (var child in newTest.ChildNamespaces) {
+							//Console.WriteLine (newTest.Title + " " + child.Title);
```
Comment (mrward) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs]: Debug code :)

**Comment by mrward** [comment]
@monojenkins rebase

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs (hunk 7 above)
Comment (ketttY) [reply main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs]: Nice catch! Thank you lol

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs [hunk 8]
```diff
@@ -268,16 +268,39 @@ void FillTests (NunitTestInfo ti)
...
-					newTest = new NUnitTestCase (this, test, test.PathName);
+				var newTest = new NUnitTestSuite (this, test);
 				newTest.FixtureTypeName = test.FixtureTypeName;
 				newTest.FixtureTypeNamespace = test.FixtureTypeNamespace;
-				Tests.Add (newTest);
 
+				bool isNameSpace = false;
+				if (test.Tests != null) {
+					foreach (NunitTestInfo child in test.Tests) {
+						if (child.Tests != null) {
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitAssemblyTestSuite.cs]: Both `isNameSpace` and `hasClassAsChild` could be computed inside the same loop:
```csharp
foreach (NunitTestInfo child in test.Tests) {
	if (child.Tests != null) {
		isNameSpace = true;

		if (child.Tests[0].Tests != null) {
			hasClassAsChild = true;
		}
		
	}
}

File: main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs [hunk 9]
```diff
@@ -80,16 +81,45 @@ protected override void OnCreateTests ()
...
+
+					bool isNameSpace = false;
+					foreach (NunitTestInfo child in test.Tests) {
+						if (child.Tests != null) {
+							isNameSpace = true;
+						}
+					}
+
+					bool hasClassAsChild = false;
+					foreach (NunitTestInfo child in test.Tests) {
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting.NUnit/MonoDevelop.UnitTesting.NUnit/NUnitTestSuite.cs]: Same as other piece of code, also this looks duplicated. Can we extract this into a method?

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests.csproj [hunk 10]
```diff
@@ -59,6 +59,10 @@
       <Name>MonoDevelop.DotNetCore</Name>
       <Private>False</Private>
     </ProjectReference>
+    <ProjectReference Include="..\..\MonoDevelop.UnitTesting.NUnit\MonoDevelop.UnitTesting.NUnit.csproj">
+      <Project>{6224D87E-2AC1-4D9F-91ED-714F797297BF}</Project>
+      <Name>MonoDevelop.UnitTesting.NUnit</Name>
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests.csproj]: No need to local copy here

Comment (Therzok) [reply main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests.csproj]:
```suggestion
      <Name>MonoDevelop.UnitTesting.NUnit</Name>
      <Private>False</Private>
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 11]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+				FixtureTypeName = "TestClassAB",
+				FixtureTypeNamespace = "A.B",
+				Name = "TestMethodAB"
+			};
+			classInfo.Tests [0] = methodInfo;
+
+			return namespaceInfo;
+		}
+
+		NunitTestInfo nestedNamespace ()
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
		NunitTestInfo NestedNamespace ()
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 12]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+				FixtureTypeName = "TestClassABC",
+				FixtureTypeNamespace = "A.B.C",
+				Name = "TestMethodABC"
+			};
+			classInfo.Tests [0] = methodInfo;
+
+			return namespaceInfo;
+		}
+
+		NunitTestInfo namespaceNoDot ()
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
		NunitTestInfo NamespaceNoDot ()
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 13]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+				FixtureTypeName = "TestClass",
+				FixtureTypeNamespace = "",
+				Name = "TestMethod"
+			};
+			classInfo.Tests [0] = methodInfo;
+
+			return classInfo;
+		}
+
+		NunitTestInfo namespaceWithDot ()
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
		NunitTestInfo NamespaceWithDot ()
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 14]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+			var case4 = new NUnitTestSuite (null, namespaceNoDot ());  
+
+			Assert.AreEqual(case1.Tests[0].Title, "TestClass");
+			Assert.AreEqual(case2.Tests[0].Title, "A.B");
+			Assert.AreEqual(case3.Tests[0].Title, "A.B.C");
+			Assert.AreEqual(case4.Tests[0].Title, "A");
+
+		}
+
+		NunitTestInfo directClassCase ()
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
		NunitTestInfo DirectClassCase ()
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 15]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+	class NUnitTestUnitTestTests : TestBase
+	{
+
+		/// <summary>
+		/// VSTS Bug 1042673: [Bug] Make the nesting for Mono NUnit test consistent with the vstests (mstest, nunit, xunit)
+		/// </summary>
+		[Test]
+		public void TestVSTS1042673 ()
+		{
+			var case1 = new NUnitTestSuite (null, directClassCase ());
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
			var case1 = new NUnitTestSuite (null, DirectClassCase ());
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 16]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+			var case2 = new NUnitTestSuite (null, namespaceWithDot ());
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
			var case2 = new NUnitTestSuite (null, NamespaceWithDot ());
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 17]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+			var case3 = new NUnitTestSuite (null, nestedNamespace ());
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
			var case3 = new NUnitTestSuite (null, NestedNamespace ());
```

File: main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs [hunk 18]
```diff
@@ -121,7 +122,141 @@ public MyVsTestUnitTest (string displayName, string fixtureTypeNamespace, string
...
+			var case4 = new NUnitTestSuite (null, namespaceNoDot ());
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.UnitTesting/MonoDevelop.UnitTesting.Tests/MonoDevelop.UnitTesting.Tests/VsTestUnitTestTests.cs]:
```suggestion
			var case4 = new NUnitTestSuite (null, NamespaceNoDot ());  
```

**Pull request merged**